packages/core/bootstrap/src/lib/ws/reducer.ts

```diff
@@ -2,61 +2,54 @@ import hash from 'object-hash'
 import { AdapterRequest } from '@chainlink/types'
 import { combineReducers, createReducer } from '@reduxjs/toolkit'
 import * as actions from './actions'
-import { WSConfig } from './types'
 import { getHashOpts } from '../util'
 
 export const getSubsId = (subscriptionMsg = {}): string => hash(subscriptionMsg, getHashOpts())
 export interface ConnectionsState {
-  /** Map of all connections by key */
-  active: {
-    [key: string]: WSConfig
-  }
-  connecting: {
-    [key: string]: number // Filter if its more than one
-  }
-  wasEverConnected: {
-    [key: string]: boolean
+  [key: string]: {
+    active: boolean
+    connecting: number
+    wasEverConnected?: boolean
   }
-  /** Number of active connections */
-  total: number
 }
 
-const initConnectionsState: ConnectionsState = { active: {}, connecting: {}, wasEverConnected: {}, total: 0 }
+const initConnectionsState: ConnectionsState = {}
 
 export const connectionsReducer = createReducer<ConnectionsState>(
   initConnectionsState,
   (builder) => {
     builder.addCase(actions.connected, (state, action) => {
       // Add connection
       const { key } = action.payload.config.connectionInfo
-      state.active[key] = action.payload.config
-      state.connecting[key] = 0
-      state.wasEverConnected[key] = true
-      // Increment num of active connections
-      state.total++
+      state[key] = {
+        active: true,
+        connecting: 0,
+        wasEverConnected: true
+      }
     })
 
     builder.addCase(actions.connect, (state, action) => {
       const { key } = action.payload.config.connectionInfo
-      const isActive = !!state.active[key]
+      const isActive = state[key]?.active
       if (isActive) return
 
-      const isConnecting = !isNaN(Number(state.connecting[key]))
-      state.connecting[key] = isConnecting ? state.connecting[key] + 1 : 1
+      const isConnecting = !isNaN(Number(state[key]?.connecting))
+      state[key] = {
+        active: false,
+        connecting: isConnecting ? state[key].connecting + 1 : 1
+      }
     })
 
     builder.addCase(actions.connectionError, (state, action) => {
-      state.connecting[action.payload.connectionInfo.key] = 0
-      delete state.active[action.payload.connectionInfo.key]
+      state[action.payload.connectionInfo.key].connecting = 0
+      state[action.payload.connectionInfo.key].active = false
     })
 
     builder.addCase(actions.disconnected, (state, action) => {
       // Remove connection
       const { key } = action.payload.config.connectionInfo
-      delete state.active[key]
-      state.connecting[key] = 0 // turn off connecting
-      // Decrement num of active connections
-      if (state.wasEverConnected[key]) state.total--
```

Comment:
We don't use it now, but we'll need to track the number of connections & subscriptions to manage the rate limits.

Reply:
I think keeping count of the subs/connections should be on the consumer side (would be safer I think too)
```js
const state = store.getState()
const activeSubscriptions = Object.values(state.ws.subscriptions).filter((s) => s.active).length
const activeConnections = Object.values(state.ws.connections).filter((c) => c.active).length
```

Reply:
It's a design choice, I'm good either way.

Reducers are here to prepare the data for consumers, but consumers can also make additional one-off aggregations if necessary. But if this starts to repeat, and it's not just one-off aggregation but starts repeating in multiple consumers, then the better move would be to bring it to the reducer directly.

Why safer though?

Reply:
For maintenance, as we are replicating the same information in two places, so we would need to maintain it in this two places. I guess not doing it properly could lead into some incongruity

Reply:
Would turning the Subscriptions and/or Connections into a class that extends the interface solve the concerns? That way you can build in the functions for filtering + returning length for tracking totals without tracking the same number in two places.

Not sure if changing this would break underlying functionality...

Reply:
See below for a potential solution that calculates the total connections state using another reducer, which provides consumers the value without needing to actively manage an additional state.